src/scp/QuorumIntersectionChecker.cpp

```diff
@@ -0,0 +1,1106 @@
+// parts we skip.
+//
+//
+// Refinement 1: Switch to checking the complement
+// ===============================================
+//
+// The nested testing and intersection loop above can be rewritten the following
+// way while preserving logical equivalence:
+//
+// foreach subset S₁ of N:
+//   if S₁ is a quorum:
+//     for each subset S₂ of N \ S₁:
+//       check that S₂ is not a quorum
+//
+// That is, any subset S₂ of the complement of S₁ is (by definition)
+// non-intersecting with S₁ and therefore needs to be not-a-quorum if we want to
+// enjoy quorum intersection. This will go a bit faster than the full cartesian
+// product space of the first algorithm because as S₁ expands, N \ S₁ contracts,
+// but the real improvement in complexity shows up if we introduce another
+// building block (which is in fact how we implement isAQuorum anyways).
+//
+//
+// Refinement 2: Switch to contracting sets to maximal quorums
+// ===========================================================
+//
+// Given some set S ⊆ N, we can contract S to a maximal quorum Q ⊆ S (if it
+// exists) in reasonably quick time: for each node Nᵢ ∈ S, remove Nᵢ from S if S
+// does not satisfy QSᵢ, and keep iterating this procedure until it reaches a
+// fixpoint. The fixpoint will either be empty (in which case S contained no
+// quorum) or the largest quorum contained within S. This runs in time linear
+// with the size of S (times at-worst linear-time slice checking, so maybe
+// quadratic; but still far better than the exponential time of the powerset!)
+//
+// Where this is useful is that the algorithm from refinement 1 can now be
+// adapted to seek quorums directly in the complement, rather than enumerating
+// subsets of the complement:
+//
+// foreach subset S₁ of N:
+//   if S₁ is a quorum:
+//     check that (N \ S₁) contracts to the empty set (i.e. has no quorum)
+//
+// This frees us from the cartesian product of the powersets. Good! But we're
+// still exploring the powerset in the outer loop.
+//
+//
+// Refinement 3: Enumerate from bottom up
+// ======================================
+//
+// We haven't mentioned yet specifically how we're going to enumerate the
+// powerset P(N), and it turns out one order is especially useful for our
+// purposes: "bottom up", starting from 1-element sets and expanding to
+// 2-element sets, then 3-element sets, and so forth.
+//
+// In particular we're going to pick a recursive enumeration that's based on two
+// sets: C (for "committed") and R (for "remaining"). We start with C = ∅ and R
+// = N. The set C at each call is the current set we're expanding, bottom-up;
+// the set R is the set we're going to draw expansions from. We pass C and R to
+// the following recursive enumerator:
+//
+//     def enumerate(C, R):
+//         return if R = ∅
+//         fail if C is a quorum and (N \ C) has a quorum
+//         pick some node Nᵢ ∈ R
+//         call enumerate(C, R-Nᵢ)
```

Comment:
if I understand correctly, for R bigger than one element, it will call line 161 again with the same parameters?

Reply:
No, it removes N_i from R on each call.

Reply:
yes, but line 161 does not depend on R at all, so it will be executed with the same parameters as before?